src/util/thread_metrics.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs;
+use std::io::{Result, Error, ErrorKind, Read};
+use std::sync::Mutex;
+use std::sync::atomic::{AtomicUsize, ATOMIC_USIZE_INIT, Ordering};
+
+use prometheus::{self, Opts, CounterVec, Desc, proto, Collector};
+
+use libc::{self, pid_t};
+
+/// monitor current process's threads.
+pub fn monitor_threads<S: Into<String>>(namespace: S) -> Result<()> {
+    let pid = unsafe { libc::getpid() };
+    let tc = try!(ThreadsColletcor::new(pid, namespace));
+    try!(prometheus::register(Box::new(tc)).map_err(|e| to_err(format!("{:?}", e))));
+
+    Ok(())
+}
+
+// Refresh monitoring threads.
+const REFRESH_CNT: usize = 10;
+
+struct ThreadsCore {
+    refresh: usize,
+    thread_ids: Vec<(pid_t, String)>,
+    cpu_totals: CounterVec,
+}
+
+struct ThreadsColletcor {
+    pid: pid_t,
+    descs: Vec<Desc>,
+    core: Mutex<ThreadsCore>,
+}
+
+impl ThreadsColletcor {
+    fn new<S: Into<String>>(pid: pid_t, namespace: S) -> Result<ThreadsColletcor> {
+        let cpu_totals = CounterVec::new(Opts::new("thread_cpu_seconds_total",
+                                                   "Total user and system CPU time spent in \
+                                                    seconds by threads.")
+                                             .namespace(namespace),
+                                         &["name"])
+            .unwrap();
+        let descs = cpu_totals.desc().into_iter().cloned().collect();
+
+        let pairs = try!(get_threads(pid));
+        let core = ThreadsCore {
+            refresh: 0,
+            thread_ids: pairs,
+            cpu_totals: cpu_totals,
+        };
+
+        Ok(ThreadsColletcor {
+            pid: pid,
+            descs: descs,
+            core: Mutex::new(core),
+        })
+    }
+}
+
+impl Collector for ThreadsColletcor {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        let mut core = self.core.lock().unwrap();
+        core.refresh += 1;
+        if core.refresh >= REFRESH_CNT {
```

Comment:
why do we need refreshing here? Getting threads every time in collect can cause performance reduction?